faiss_vector_cache.go

```diff
@@ -52,34 +52,83 @@ func (vc *vectorIndexCache) Clear() {
 	vc.m.Unlock()
 }
 
-func (vc *vectorIndexCache) loadOrCreate(fieldID uint16, mem []byte, except *roaring.Bitmap) (
-	index *faiss.IndexImpl, vecDocIDMap map[int64]uint32, vecIDsToExclude []int64, err error) {
-	var found bool
-	index, vecDocIDMap, vecIDsToExclude, found = vc.loadFromCache(fieldID, except)
-	if !found {
-		index, vecDocIDMap, vecIDsToExclude, err = vc.createAndCache(fieldID, mem, except)
-	}
-	return index, vecDocIDMap, vecIDsToExclude, err
+// loadDocVecIDMap indicates if a non-nil docVecIDMap should be returned.
+// It is true when a filtered kNN query accesses the cache since it requires the
+// map. It's false otherwise.
+func (vc *vectorIndexCache) loadOrCreate(fieldID uint16, mem []byte,
+	loadDocVecIDMap bool, except *roaring.Bitmap) (
+	index *faiss.IndexImpl, vecDocIDMap map[int64]uint32, docVecIDMap map[uint32]int64,
+	vecIDsToExclude []int64, err error) {
+	index, vecDocIDMap, docVecIDMap, vecIDsToExclude, err = vc.loadFromCache(
+		fieldID, loadDocVecIDMap, mem, except)
+	return index, vecDocIDMap, docVecIDMap, vecIDsToExclude, err
 }
 
-func (vc *vectorIndexCache) loadFromCache(fieldID uint16, except *roaring.Bitmap) (
-	index *faiss.IndexImpl, vecDocIDMap map[int64]uint32, vecIDsToExclude []int64, found bool) {
+// function to load the vectorDocIDMap and if required, docVecIDMap from cache
+// if possible.
+// If not, it will create these and add them to the cache.
+func (vc *vectorIndexCache) loadFromCache(fieldID uint16, loadDocVecIDMap bool,
+	mem []byte, except *roaring.Bitmap) (index *faiss.IndexImpl, vecDocIDMap map[int64]uint32,
+	docVecIDMap map[uint32]int64, vecIDsToExclude []int64, err error) {
+
 	vc.m.RLock()
 	defer vc.m.RUnlock()
 
 	entry, ok := vc.cache[fieldID]
-	if !ok {
-		return nil, nil, nil, false
+	// either do not require the docVecIDMap or if we do, it's already loaded in
+	// the cache.
+	if ok && (!loadDocVecIDMap || (loadDocVecIDMap && entry.docVecIDMap != nil)) {
+		index, vecDocIDMap, docVecIDMap = entry.load()
+		vecIDsToExclude = getVecIDsToExclude(vecDocIDMap, except)
+		return index, vecDocIDMap, docVecIDMap, vecIDsToExclude, nil
```

Comment:
RUnlock's context needs to end here, and a write lock is needed for everything that follows. You can refactor the existing createAndCache to createAndCacheLOCKED to achieve this and it seems you must check for loadDocVecIDMap within right lock again unfortunately.

Reply:
my bad, missed it the first time.
I've changed that in my recent commits - modifying the cache is done only after acquiring a lock, reading without modifying done under a read lock.